src/providers/crypto_aescbc.rs

```diff
@@ -0,0 +1,592 @@
+    /// ```
+    ///
+    /// It is **required** that the specified initialization vector be at
+    /// maximum 16 bytes long. If you give us a value that is longer than 16
+    /// bytes, it will be clipped or filled with zeros if it is shorter.
+    ///
+    /// [`AesCbcIv::Custom`]: ./enum.AesCbcIv.html#variant.Custom
+    fn from(value: String) -> Self {
+        value.as_str().into()
+    }
+}
+
+/// A crypto that uses the AES encryption algorithm with CBC mode.
+///
+/// Crypto provides interface fore data encryption and decryption.
+///
+/// # Examples
+/// Create cryptor with cipher key and random initialization vector.
+/// ```rust
+/// # use pubnub::{
+/// #     core::{error::PubNubError, Cryptor},
+/// #     providers::crypto_aescbc::{AesCbcIv, AesCbcCrypto}
+/// # };
+/// #
+/// # fn main() -> Result<(), PubNubError> {
+/// let cryptor = AesCbcCrypto::new("enigma", AesCbcIv::Random)?;
+/// # Ok(())
+/// # }
+/// ```
+pub struct AesCbcCrypto {
+    cipher_key: Vec<u8>,
+    iv: Option<Vec<u8>>,
+    iv_constant: bool,
+}
+
+impl AesCbcCrypto {
+    /// Create an AES-CBC crypto for data encryption and decryption.
+    ///
+    /// It is **required** that the specified initialization vector be at
+    /// maximum 16 bytes long. If you give us a value that is longer than 16
+    /// bytes, it will be clipped or filled with zeros if it is shorter.
+    ///
+    /// # Errors
+    /// Should return an [`PubNubError::CryptoInitializationError`] if cipher
+    /// key or initialization vectors are empty.
+    pub fn new<C, I>(cipher_key: C, iv: I) -> Result<Self, PubNubError>
+    where
+        C: Into<Vec<u8>>,
+        I: Into<AesCbcIv>,
+    {
+        let cipher_key: Vec<u8> = cipher_key.into();
+        let vector: AesCbcIv = iv.into();
+        let iv_constant = matches!(vector, AesCbcIv::Constant);
+
+        if let AesCbcIv::Custom(custom_iv) = &vector {
+            if custom_iv.is_empty() {
+                return Err(PubNubError::CryptoInitializationError(
+                    "Initialization vector is empty".into(),
+                ));
+            }
+        }
+
+        if cipher_key.is_empty() {
+            return Err(PubNubError::CryptoInitializationError(
+                "Cipher key is empty".into(),
+            ));
+        }
+
+        Ok(AesCbcCrypto {
+            cipher_key: AesCbcCrypto::sha256_hex(cipher_key),
+            iv: match vector {
+                AesCbcIv::Constant => Some(b"0123456789012345".to_vec()),
+                AesCbcIv::Custom(custom) => Some(custom),
+                AesCbcIv::Random => None,
+            },
+            iv_constant,
+        })
+    }
+
+    /// Calculate size of buffer for processed data.
+    ///
+    /// Buffers may require different sizes of allocated memory depending on
+    /// operation and initialization vector type.
+    fn estimated_buffer_size(&self, source: &[u8], encryption: bool) -> usize {
```

Comment:
Yes, sure. So when you use it you need to know exactly which version you want to have, either it's encryption or decryption. Why make this decision second time in code by examining this `encryption` boolean? The methods are simpler, you don't have this boolean at all, seems like a win win

```
fn estimated_enc_buffer_size(&self, source: &[u8]) {
        let mut size = source.len();

        // More space required for encryption to align with AES cipher block
        // size.
        size += (AES_BLOCK_SIZE - size % AES_BLOCK_SIZE) + AES_BLOCK_SIZE;
        // Reserve more space to store random initialization vector.
        if !self.iv_constant {
           size += AES_BLOCK_SIZE;
        }
        size
}

fn estimated_dec_buffer_size(&self, source: &[u8]) {
      let mut size = source.len();
      if !self.iv_constant {
           size += AES_BLOCK_SIZE;
      }
      size
}

```

Reply:
It will be worth if we want to use the `optional` syntax. Then the codes will be short, redable and without mutability.

Reply:
@Xavrax Not sure how does this apply to this discussion. Care to explain it a little further?

Reply:
@kleewho @Xavrax I can suggest this variant, but it looses comments:
```rust
fn estimated_enc_buffer_size(&self, source: &[u8]) -> usize {
    source.len()
        + AES_BLOCK_SIZE
        + (AES_BLOCK_SIZE - source.len() % AES_BLOCK_SIZE)
        + if !self.iv_constant { AES_BLOCK_SIZE } else { 0 }
}

fn estimated_dec_buffer_size(&self, source: &[u8]) -> usize {
    source.len() - if !self.iv_constant { AES_BLOCK_SIZE } else { 0 }
}
```

Reply:
We had a discussion on this method above, where I left my response. Let it is being as is because there are no clear signs of considerable improvements (at least for me). That flag at least will force you to think in which context method used, when named method can be added into wrong place.

Reply:
> @Xavrax Not sure how does this apply to this discussion. Care to explain it a little further?

I though about the optional syntax if we split that buffers. 

But imo @kleewho has right about the readablility. 
Keep what we have but with the separation.

Reply:
@Xavrax I've mentioned above, that that flag increases awareness of where method used :)

@kleewho @Xavrax as reviewers tell me then, should I split it into two methods or not.

Reply:
Imo that would be nice.

Reply:
@kleewho I've pushed split implementation. If it is OK and the rest of the changes are fine, we can merge it.